internal/controller/state/graph/route_common.go

```diff
@@ -1150,3 +1233,168 @@ func routeKeyForKind(kind v1.Kind, nsname types.NamespacedName) RouteKey {
+	if len(config.rules) == 0 {
+		r.Valid = false
+		cond := conditions.NewRouteBackendRefUnsupportedValue(
+			"Must have at least one Rule",
+		)
+		r.Conditions = append(r.Conditions, cond)
+		return r
+	}
+
+	// Process all BackendRefs from all rules
+	var allBackendRefs []BackendRef
+	var allConditions []conditions.Condition
+
+	for ruleIdx, rule := range config.rules {
+		if len(rule.backendRefs) == 0 {
+			continue
+		}
+
+		for refIdx, ref := range rule.backendRefs {
+			br, conds := validateBackendRefL4RouteMulti(
+				config.namespace, ref, services, r.ParentRefs,
+				config.refGrantResolver, ruleIdx, refIdx,
+			)
+			allBackendRefs = append(allBackendRefs, br)
+			allConditions = append(allConditions, conds...)
+		}
+	}
+
+	// Set BackendRefs for multi-backend support
+	r.Spec.BackendRefs = allBackendRefs
+	r.Valid = true
+	r.Attachable = true
+
+	if len(allConditions) > 0 {
+		r.Conditions = append(r.Conditions, allConditions...)
+	}
+
+	return r
+}
+
+// validateBackendRefL4RouteMulti is a generic function to validate BackendRef for both TCP and UDP routes.
+// This eliminates code duplication between validateBackendRefTCPRouteMulti and validateBackendRefUDPRouteMulti.
+func validateBackendRefL4RouteMulti(
+	namespace string,
+	ref v1alpha.BackendRef,
+	services map[types.NamespacedName]*apiv1.Service,
+	parentRefs []ParentRef,
+	refGrantResolver func(resource toResource) bool,
+	ruleIdx int,
+	refIdx int,
+) (BackendRef, []conditions.Condition) {
+	refPath := field.NewPath("spec").Child("rules").Index(ruleIdx).Child("backendRefs").Index(refIdx)
+
+	if valid, cond := validateBackendRef(
+		ref,
+		namespace,
+		refGrantResolver,
+		refPath,
+	); !valid {
+		backendRef := BackendRef{
+			Valid: false,
+		}
+
+		return backendRef, []conditions.Condition{cond}
+	}
+
+	ns := namespace
+	if ref.Namespace != nil {
+		ns = string(*ref.Namespace)
+	}
+
+	svcNsName := types.NamespacedName{
+		Namespace: ns,
+		Name:      string(ref.Name),
+	}
+
+	svcIPFamily, svcPort, err := getIPFamilyAndPortFromRef(
+		ref,
+		svcNsName,
+		services,
+		refPath,
+	)
+
+	// Handle weight - default to 1 if not specified
+	weight := int32(1)
+	if ref.Weight != nil {
+		if validateWeight(*ref.Weight) != nil {
+			weight = 0 // Invalid weight set to 0 (no traffic)
+		} else {
+			weight = *ref.Weight
+		}
+	}
+
+	backendRef := BackendRef{
+		SvcNsName:   svcNsName,
+		ServicePort: svcPort,
+		Weight:      weight,
+		Valid:       true,
+	}
+
+	if err != nil {
+		backendRef.Valid = false
+
+		return backendRef, []conditions.Condition{conditions.NewRouteBackendRefRefBackendNotFound(err.Error())}
+	}
+
+	// For TCP/UDPRoute, we don't need to validate app protocol compatibility
+	// as TCP/UDP are protocol-agnostic at the application layer
+
+	var conds []conditions.Condition
+	for _, parentRef := range parentRefs {
+		if err := verifyIPFamily(parentRef.Gateway.EffectiveNginxProxy, svcIPFamily); err != nil {
+			backendRef.Valid = backendRef.Valid || false
```

Comment:
This is confusing and looks incorrect. Shouldn't it just be 
```suggestion
			backendRef.Valid = false
```
?